File: ocaml/p0001.ml
(* Project Euler: Problem 1 *)

let make_seq upper =
  List.filter (fun n -> n mod 3 = 0 || n mod 5 = 0) (List.init upper (fun n -> n));;

let rec sum_of_seq l =
  match l with
  | [] -> 0
  | h :: rest -> h + (sum_of_seq rest);;

let () =
  Printf.printf "sum of all the multiples of 3 or 5 below 1000 is %d\n" (sum_of_seq (make_seq 1000));;

File: rust/src/bin/p0001.rs
// Project Euler: Problem 1

euler::run_solver!(1);

fn solve() -> String {
    compute(1000).to_string()
}

fn compute(limit: i64) -> i64 {
    debug_assert!(limit > 0);

    let f = |x: i64| -> i64 {
        // tmp: number of multiples of `x` less than `limit`
        let tmp = (limit - 1) / x;
        (1 + tmp) * tmp / 2 * x
    };

    f(3) + f(5) - f(15)
}

#[cfg(test)]
mod tests {
    use super::compute;

    #[test]
    fn p0001_sum_less_than_10() {
        assert_eq!(compute(10), 23);
    }

    #[test]
    fn p0001_sum_less_than_1000() {
        assert_eq!(compute(1000), 233168);
    }
}
